controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go

```diff
@@ -234,7 +235,10 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
 	// Upgrade takes precedence over other operations
 	if len(requireUpgrade) > 0 {
```

Comment:
This is actually going to detect changes in KubeadmConfigSpec too, it seems? I thought we weren't gunna try to handle that. 

I want to though, so I'm down. :D

Reply:
> This is actually going to detect changes in KubeadmConfigSpec too, it seems? I thought we weren't gunna try to handle that.

It shouldn't, it's comparing based on the computed hash, which only takes into account Version and InfrastructureTemplate fields: https://github.com/kubernetes-sigs/cluster-api/blob/5aea6d2b61ed0384d2289e248af08293f74e079e/controlplane/kubeadm/internal/hash/hash.go#L34-L49

Reply:
dang it!!! i didn't dig deep enough. my dreams are crushed. 

What would actually happen though if it considered the kubeadmconfigspec? I think I'm missing the big problem.

Reply:
More specifically, `KubeadmConfigSpec.JoinConfiguration`.

Reply:
I suppose the worst thing that would happen is someone b0rks their cluster, but since we are blocking all operations on control plane/etcd health, the blast radius should be relatively small.

Reply:
I raised this here: https://github.com/kubernetes-sigs/cluster-api/issues/2083 and thought that there may be some cases i'm missing.

I bring it up because for us, machine replacement is the right mechanism for updating flags, and it's also the right mechanism for version upgrades. So why not include both cases in the same flow?

Reply:
@ncdc @vincepri @chuckha What do you think about allowing mutation of JoinConfiguration? It should be easy enough to patch up as part of this PR and I believe the safeguards we are putting in place should be enough to limit the blast radius of the footgun.

Reply:
Totally cool for it to be follow-up work, and happy to do testing. I just wanted to see if it's something we'd be okay with.